examples/charm++/future/fib.C

```diff
@@ -1,40 +1,79 @@
+#include <cstdint>
+using fibonacci_t = std::uint64_t;
+
 #include "fib.decl.h"
 
 #define THRESHOLD 3
 
 struct Main : public CBase_Main {
-  Main(CkArgMsg *m) { thisProxy.run(atoi(m->argv[1])); }
+  Main(CkArgMsg *m) {
+    if (m->argc > 1) thisProxy.run(atoi(m->argv[1]));
+    else thisProxy.run(24);
+  }
 
   void run(int n) {
-    ck::future<int> f;
+    ck::future<fibonacci_t> f;
     CProxy_Fib::ckNew(n, f);
-    CkPrintf("Fibonacci number is: %d\n", f.get());
+    ckout << "Fibonacci number is: " << f.get() << endl;
     f.release();
     CkExit();
   }
 };
 
 struct Fib : public CBase_Fib {
-  Fib(int n, const ck::future<int> &prev_) : prev(prev_) { thisProxy.calc(n); }
+  Fib(int n, const ck::future<fibonacci_t> &prev_) : prev(prev_) { thisProxy.calc(n); }
 
   int seqFib(int n) { return (n < 2) ? n : seqFib(n - 1) + seqFib(n - 2); }
 
   void calc(int n) {
     if (n < THRESHOLD) {
       prev.set(seqFib(n));
     } else {
-      ck::future<int> f1, f2;
-      CProxy_Fib::ckNew(n - 1, f1);
-      CProxy_Fib::ckNew(n - 2, f2);
-      prev.set(f1.get() + f2.get());
-      f1.release();
-      f2.release();
+      std::vector<ck::future<fibonacci_t>> pending(2);
+      CProxy_Fib::ckNew(n - 1, pending[0]);
+      CProxy_Fib::ckNew(n - 2, pending[1]);
+      fibonacci_t sum = 0;
+      if (n % 2 == 0) {
+        // even n's use wait any
+        while (!pending.empty()) {
+          // wait for any of the futures in pending to become ready
+          auto pair = ck::wait_any(pending.begin(), pending.end());
+          // add the received value to the sum
+          sum += pair.first;
+          // then release and erase the fulfilled future
+          pair.second->release();
+          pending.erase(pair.second);
+        }
+      } else if (n % 3 == 0) {
+        // div by three (but not by two) use wait_some
+        auto pair = ck::wait_some(pending.begin(), pending.end());
+        for (const auto& value : pair.first) {
+          sum += value;
+        }
+        // in a real program, we could go off and do something else...
+        // but we'll just wait on whatever's still outstanding
+        for (auto it = pair.second; it < pending.end(); it++) {
+          sum += it->get();
+        }
+      } else {
+        // odds use wait all
```

Comment:
Nitpicking, but I'd change this comment to make it clear that this branch applies to odds that are not divisible by three.